### ISSUE TITLE
Non-deterministic hang in quic.Dial when PacketConn returns errors

### ISSUE THREAD

**Issue by Eudi4H** [lucas-clemente/quic-go#2172]
I'm using quic-go v0.12.1 with go1.12.10. When using a `PacketConn` that returns errors, `quic.Dial` may not return. For me this happens with a probability of about 1%. Here is a small program that reproduces it.

<details>
<summary>go.mod</summary>

<pre>
module example.com/example

require github.com/lucas-clemente/quic-go v0.12.1
</pre>
</details>

<details>
<summary>demo.go</summary>

<pre>
// Demonstrates a hang in quic-go v0.12.1 when the underlying PacketConn returns
// errors during quic.Dial. Usage:
//   ./demo
// The program will count the number of connection attempts that completed
// before hitting one that hangs. It seems more likely to hang when there is a
// slight delay before ReadFrom returns with an error. Try this:
//   ./demo -read-delay 800us

package main

import (
	"crypto/tls"
	"flag"
	"fmt"
	"net"
	"time"

	"github.com/lucas-clemente/quic-go"
)

// stringAddr implements the net.Addr interface by returning static strings.
type stringAddr struct{ network, address string }

func (addr stringAddr) Network() string {
	return addr.network
}

func (addr stringAddr) String() string {
	return addr.address
}

var errNotImplemented error = fmt.Errorf("not implemented")

// brokenPacketConn returns errNotImplemented for all its methods. readDelay is
// a delay to insert before returning from ReadFrom.
type brokenPacketConn struct {
	readDelay time.Duration
}

func (c *brokenPacketConn) ReadFrom(p []byte) (int, net.Addr, error) {
	// fmt.Printf("ReadFrom attempt\n")
	// defer fmt.Printf("ReadFrom return\n")
	time.Sleep(c.readDelay)
	return 0, nil, errNotImplemented
}

func (c *brokenPacketConn) WriteTo(p []byte, addr net.Addr) (int, error) {
	// fmt.Printf("WriteTo attempt\n")
	// defer fmt.Printf("WriteTo return\n")
	return 0, errNotImplemented
}

func (c *brokenPacketConn) Close() error {
	return errNotImplemented
}

func (c *brokenPacketConn) LocalAddr() net.Addr {
	return stringAddr{"local", "local"}
}

func (c *brokenPacketConn) SetDeadline(t time.Time) error {
	return errNotImplemented
}

func (c *brokenPacketConn) SetReadDeadline(t time.Time) error {
	return errNotImplemented
}

func (c *brokenPacketConn) SetWriteDeadline(t time.Time) error {
	return errNotImplemented
}

func run(readDelay time.Duration) {
	pconn := &brokenPacketConn{readDelay}
	tlsConfig := &tls.Config{
		InsecureSkipVerify: true,
		NextProtos:         []string{"quicdemo"},
	}
	quicConfig := &quic.Config{}
	// fmt.Printf("before quic.Dial\n")
	sess, err := quic.Dial(pconn, nil, "", tlsConfig, quicConfig)
	// fmt.Printf("after quic.Dial %v %v\n", sess, err)
	if err == nil {
		sess.Close()
	}
}

func main() {
	readDelay := flag.Duration("read-delay", 0*time.Second, "delay before returning from ReadFrom")
	flag.Parse()
	for i := 0; ; i++ {
		fmt.Printf("%d\n", i)
		run(*readDelay)
	}
}
</pre>
</details>

The program will count the number of `quic.Dial`s that return before finding one that hangs. For me, the output is like this:
```
$ ./demo
0
[...]
38
39
40

$ ./demo
0
[...]
146
147
148

$ ./demo
0
[...]
110
111
112
```

The probability of a hang is increased if there's a short delay before `ReadFrom` returns with its error. (In my original use case, this was being caused by a `net.Dial` to a closed port.) For me, a delay of about 800 μs works pretty well to reproduce the hang quickly.

```
$ ./demo -read-delay 800us
0
1
2

$ ./demo -read-delay 800us
0
1
2

$ ./demo -read-delay 800us
0
```

Another easy way to reproduce the hang is with a closed `UDPConn`:

<details>
<summary>demo.go using a closed net.UDPConn</summary>

<pre>
// Demonstrates a hang in quic-go v0.12.1 when the underlying PacketConn returns
// errors during quic.Dial. Usage:
//   ./demo
// The program will count the number of connection attempts that completed
// before hitting one that hangs.

package main

import (
	"crypto/tls"
	"fmt"
	"net"

	"github.com/lucas-clemente/quic-go"
)

func run() {
	pconn, err := net.ListenPacket("udp", "127.0.0.1:8888")
	if err != nil {
		panic(err)
	}
	pconn.Close()
	tlsConfig := &tls.Config{
		InsecureSkipVerify: true,
		NextProtos:         []string{"quicdemo"},
	}
	quicConfig := &quic.Config{}
	// fmt.Printf("before quic.Dial\n")
	sess, err := quic.Dial(pconn, nil, "", tlsConfig, quicConfig)
	// fmt.Printf("after quic.Dial %v %v\n", sess, err)
	if err == nil {
		sess.Close()
	}
}

func main() {
	for i := 0; ; i++ {
		fmt.Printf("%d\n", i)
		run()
	}
}
</pre>
</details>

**Comment by marten-seemann** [comment]
I can't reproduce this using your `demo.go` and the current master branch.

**Comment by Eudi4H** [comment]
> I can't reproduce this using your `demo.go` and the current master branch.

You're right. I tried again using 3ea7ad19 and it doesn't occur.

I bisected and found that the problem was fixed by d689f9a3 (#2097/#2098).
